types_test.go

```diff
@@ -6397,3 +6398,181 @@ func TestMissingObjectFields(t *testing.T) {
+		Str                     string
+		OptStr                  OptionalStr
+		OptStrNull              OptionalStr
+		Bool                    bool
+		OptBool                 OptionalBool
+		OptBoolNull             OptionalBool
+		Int16                   int16
+		OptInt16                OptionalInt16
+		OptInt16Null            OptionalInt16
+		Int32                   int32
+		OptInt32                OptionalInt32
+		OptInt32Null            OptionalInt32
+		Int64                   int64
+		OptInt64                OptionalInt64
+		OptInt64Null            OptionalInt64
+		Float32                 float32
+		OptFloat32              OptionalFloat32
+		OptFloat32Null          OptionalFloat32
+		Float64                 float64
+		OptFloat64              OptionalFloat64
+		OptFloat64Null          OptionalFloat64
+		BigInt                  *big.Int
+		OptBigInt               OptionalBigInt
+		OptBigIntNull           OptionalBigInt
+		Uuid                    UUID
+		OptUuid                 OptionalUUID
+		OptUuidNull             OptionalUUID
+		Bytes                   []byte
+		OptBytes                OptionalBytes
+		OptBytesNull            OptionalBytes
+		DateTime                time.Time
+		OptDateTime             OptionalDateTime
+		OptDateTimeNull         OptionalDateTime
+		LocalDateTime           LocalDateTime
+		OptLocalDateTime        OptionalLocalDateTime
+		OptLocalDateTimeNull    OptionalLocalDateTime
+		LocalTime               LocalTime
+		OptLocalTime            OptionalLocalTime
+		OptLocalTimeNull        OptionalLocalTime
+		LocalDate               LocalDate
+		OptLocalDate            OptionalLocalDate
+		OptLocalDateNull        OptionalLocalDate
+		Duration                Duration
+		OptDuration             OptionalDuration
+		OptDurationNull         OptionalDuration
+		RelativeDuration        RelativeDuration
+		OptRelativeDuration     OptionalRelativeDuration
+		OptRelativeDurationNull OptionalRelativeDuration
+	}
+
+	testJson := []byte(`{
+		"Str": "test str",
+		"OptStr": "null test str",
+		"OptStrNull": null,
+		"Bool": true,
+		"OptBool": true,
+		"OptBoolNull": null,
+		"Int16": 12345,
+		"OptInt16": 12345,
+		"OptInt16Null": null,
+		"Int32": 12345,
+		"OptInt32": 12345,
+		"OptInt32Null": null,
+		"Int64": 12345,
+		"OptInt64": 12345,
+		"OptInt64Null": null,
+		"Float32": 12345,
+		"OptFloat32": 12345,
+		"OptFloat32Null": null,
+		"Float64": 12345,
+		"OptFloat64": 12345,
+		"OptFloat64Null": null,
+		"BigInt": 123456789012345678901234567890,
+		"OptBigInt": 123456789012345678901234567890,
+		"OptBigIntNull": null,
+		"Uuid": "759637d8-6635-11e9-b9d4-098002d459d5",
+		"OptUuid": "759637d8-6635-11e9-b9d4-098002d459d5",
+		"OptUuidNull": null,
+		"Bytes": "cXdlcnR5Cgl1aW9w",
+		"OptBytes": "cXdlcnR5Cgl1aW9w",
+		"OptBytesNull": null,
+		"DateTime": "2021-10-01T12:34:56.123456789Z",
+		"OptDateTime": "2021-10-01T12:34:56.123456789Z",
+		"OptDateTimeNull": null,
+		"LocalDateTime": {},
+		"OptLocalDateTime": {},
```

Comment:
I don't think this implementation is a good idea. What good is serializing a local date time (or any of the other `cal::local_*`s) as `{}`?

Reply:
I was trying to make all the optional types behave consistently, so they all behave the same as `json.Marshal` called on their inner value if it's set. But you're right, currently serializing the local datetime types isn't very useful, so maybe we should add marshallers to those types that just return/parse the string representation in the same way edgedb itself does:
```
edgedb> select <json>{
.......   localdatetime := cal::to_local_datetime(2021, 10, 1, 12, 34, 56),
.......   localdate := cal::to_local_date(2021, 10, 1)
....... };
{"{\"localdate\": \"2021-10-01\", \"localdatetime\": \"2021-10-01T12:34:56\"}"}
```

Reply:
Sounds good to me.